File: protocol/bridge/common/src/main/java/ink/reactor/protocol/bridge/common/handler/ArrayHandlerStorage.java
package ink.reactor.protocol.bridge.common.handler;

import ink.reactor.protocol.api.PlayerConnection;
import ink.reactor.protocol.api.buffer.reader.ReaderBuffer;
import ink.reactor.protocol.api.packet.PacketHandler;

import java.util.Arrays;

public final class ArrayHandlerStorage implements MutablePacketHandlerStorage {
    private PacketHandler[] handlers;

    public ArrayHandlerStorage(PacketHandler[] handlers) {
        this.handlers = handlers;
    }

    @Override
    public void callHandlers(final PlayerConnection connection, final ReaderBuffer readerBuffer) {
        for (final PacketHandler handler : handlers) {
            handler.handle(connection, readerBuffer);
        }
    }

    @Override
    public MutablePacketHandlerStorage add(final PacketHandler handler) {
        handlers = Arrays.copyOf(handlers, handlers.length+1);
        handlers[handlers.length - 1] = handler;
        return this;
    }

    @Override
    public MutablePacketHandlerStorage remove(final PacketHandler handler) {
        int index = -1;
        for (int i = 0; i < handlers.length; i++) {
            if (handler.equals(handlers[i])) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            return this;
        }
        if (handlers.length == 2) {
            return new SingleHandlerStorage(handlers[index == 0 ? 1 : 0]);
        }
        if (handlers.length == 1) {
            return null;
        }

        final PacketHandler[] newHandlers = new PacketHandler[handlers.length - 1];

        System.arraycopy(handlers, 0, newHandlers, 0, index);
        System.arraycopy(handlers, index + 1, newHandlers, index, handlers.length - index - 1);

        this.handlers = newHandlers;
        return this;
    }

    @Override
    public MutablePacketHandlerStorage add(final PacketHandler... handlers) {
        final PacketHandler[] newHandlers = Arrays.copyOf(this.handlers, this.handlers.length + handlers.length);
        System.arraycopy(handlers, 0, newHandlers, this.handlers.length, handlers.length);
        this.handlers = newHandlers;
        return this;
    }

    @Override
    public MutablePacketHandlerStorage remove(final PacketHandler... handlers) {
        MutablePacketHandlerStorage storage = this;
        for (final PacketHandler handler : handlers) {
            storage = remove(handler);
        }
        return storage;
    }
}

File: networking/internal/src/main/kotlin/ink/reactor/network/internal/packet/storage/ArrayHandlerStorage.kt
package ink.reactor.network.internal.packet.storage

import ink.reactor.network.api.packet.Packet
import ink.reactor.network.api.packet.handler.PacketHandler
import ink.reactor.network.api.player.PlayerConnection

internal class ArrayHandlerStorage(private var handlers: Array<PacketHandler>) : MutablePacketHandlerStorage {

    override fun callHandlers(connection: PlayerConnection, packet: Packet) {
        for (handler in handlers) {
            handler.handle(connection, packet)
        }
    }

    override fun add(handler: PacketHandler): MutablePacketHandlerStorage {
        handlers = handlers.plus(handler)
        return this
    }

    override fun remove(handler: PacketHandler): MutablePacketHandlerStorage? {
        val index = handlers.indexOf(handler)
        if (index == -1) return this

        return when (handlers.size) {
            1 -> null
            2 -> SingleHandlerStorage(handlers[if (index == 0) 1 else 0])
            else -> {
                val newHandlers = handlers.toMutableList().apply { removeAt(index) }.toTypedArray()
                this.handlers = newHandlers
                this
            }
        }
    }

    override fun add(vararg handlers: PacketHandler): MutablePacketHandlerStorage {
        this.handlers = this.handlers.plus(handlers)
        return this
    }

    override fun remove(vararg handlers: PacketHandler): MutablePacketHandlerStorage? {
        var storage: MutablePacketHandlerStorage? = this
        for (handler in handlers) {
            storage = storage?.remove(handler)
        }
        return storage
    }
}
